### PR TITLE
chat-js: fix scrolling and unread behaviour

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
- Fixes autoscroll to the bottom on new messages
- Changes behaviour of unread markers: The chat loads in centered at the unread marker, and sets the chat as read when the user scrolls to the bottom. This causes the 'x new messages since y' overlay to disappear, however the green line break does not disappear until the user navigates away and back to the chat. Additionally, if all the messages in a chat have been read and a new message comes in whilst the chat is open, the new message is automatically marked as read
- Hides line on left side of 'new messages below' marker on small screens
- Passes prop.size into the flexBasis property of the sigil, so that it flexes correctly

cc: @urcades 

fixes: #2778
fixes: #2782
fixes: #2797
fixes: #2804
fixes: #2817

### CODE REVIEW

**Comment by matildepark** [comment]
Does this address complaints 2 and 3 of #2817?

**Comment by matildepark** [comment]
Also how does this address #2778? It's pretty widespread on livenet for repros ofc. Does this fix the "pin chat to bottom" behaviour? Just confused as to what's going on for our scroll lock behaviour here.

**Comment by loganallenc** [comment]
What happens when the unread marker is more than 220 messages old? (In other words, further back than a single fetch of backlog messages would give us.

File: pkg/interface/chat/src/js/components/chat.js [hunk 1]
```diff
@@ -89,21 +96,21 @@ export class ChatScreen extends Component {
...
        props.envelopes.length !== prevProps.envelopes.length ||
        getNumPending(props) !== this.lastNumPending ||
        state.numPages !== prevState.numPages)
     ) {
-      if(state.scrollBottom) {
-        setTimeout(() => {
-          this.scrollToBottom();
-        })
-      } else {
+      this.scrollToBottom();
```
Comment (liam-fitzgerald) [inline pkg/interface/chat/src/js/components/chat.js:112]: here we try to scrollToBottom upon loading any new messages. Safari not scrolling to bottom on load is a result of a race condition where we try to scroll to the bottom before any messages load in. 
@matildepark

**Comment by liam-fitzgerald** [comment]
> What happens when the unread marker is more than 220 messages old? (In other words, further back than a single fetch of backlog messages would give us.

Good catch. On one hand I'd like to fetch all unread messages, but I wonder if this is feasible. What's your take?

**Comment by liam-fitzgerald** [comment]
> Does this address complaints 2 and 3 of #2817?

yes

**Comment by urcades** [comment]
@matildepark + @liam-fitzgerald 

I sent this to Liam this morning, but figured it'd be well to just post it so everyone is on the same page with my thinking:

Screen Shot 2020-04-29 at 11 36 27 AM

The tweet I reference in particular is [here](https://twitter.com/rsnous/status/1251810672544845826?s=20), and it rings deeply true to me — in many platforms, twitter being the most egregious one, scrolling is often a "real labor" labor performed with no respect being given to it by the platform as such. 

Ideally, in -all- cases, no matter what, I'd heavily prefer we do whatever we can to respect our user's scrolling labor and never take it away from them — the only times we should automatically progress the scroll is when they are already situated near the bottom of the chat, with a small threshold of what constitutes "near the bottom".

The example I provided Liam is that perhaps the height of a single line of text, or an avatar, or something in that realm (the max of this threshold perhaps being half the height of the viewport) is enough to keep a users "pinned to the bottom". We should assume that any distance further than this constitutes a deliberate act to stay situated in the scroll wherever they are, and respect that positioning.

The only other case in which we'd auto-scroll a user is if they input text into the messaging-send-field, and send it, or post an image, or code, or whatever else we allow a user to send in the future — posting to the mainline view of a chat should assume that one is responding to the latest messages, which are always located at the bottom of a scroll.

Let me know if these thoughts seem sane/amenable to how Chat is currently built — I have no idea how to address the case Logan mentioned, regarding fetching "further than the backlog history". My naive answer to this would involve loading in whatever history is necessary to situate a user where they last left off, but I'm unsure if that's possible or ergonomic or involves longer load/wait times.

**Comment by loganallenc** [comment]
> > What happens when the unread marker is more than 220 messages old? (In other words, further back than a single fetch of backlog messages would give us.
> 
> 
> 
> Good catch. On one hand I'd like to fetch all unread messages, but I wonder if this is feasible. What's your take?

I doubt it would be slow to do so. Go for it!

**Comment by loganallenc** [comment]
I’m not a fan of doing this in the askForMessages method. I’d rather a method that wraps that method or calls that method.

**Comment by loganallenc** [comment]
King, looks good. I’ll test it tomorrow.

**Comment by liam-fitzgerald** [comment]
Ping @loganallenc

**Pull request merged**